crates/turborepo-lib/src/daemon/client.rs

```diff
@@ -0,0 +1,134 @@
+            .daemon_status
+            .ok_or(DaemonError::MissingResponse)
+    }
+
+    /// Stops the daemon and closes the connection, returning
+    /// the connection settings that were used to connect.
+    pub async fn stop(mut self) -> Result<T, DaemonError> {
+        self.client.shutdown(proto::ShutdownRequest {}).await?;
+        Ok(self.connect_settings)
+    }
+
+    /// Interrogate the server for its version.
+    pub(super) async fn handshake(&mut self) -> Result<(), DaemonError> {
+        let _ret = self
+            .client
+            .hello(proto::HelloRequest {
+                version: get_version().to_string(),
+                // todo(arlyon): add session id
+                ..Default::default()
+            })
+            .await?;
+
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    pub async fn get_changed_outputs(
+        &mut self,
+        hash: String,
+        output_globs: Vec<String>,
+    ) -> Result<Vec<String>, DaemonError> {
+        Ok(self
+            .client
+            .get_changed_outputs(proto::GetChangedOutputsRequest { hash, output_globs })
+            .await?
+            .into_inner()
+            .changed_output_globs)
+    }
+
+    #[allow(dead_code)]
+    pub async fn notify_outputs_written(
+        &mut self,
+        hash: String,
+        output_globs: Vec<String>,
+        output_exclusion_globs: Vec<String>,
+    ) -> Result<(), DaemonError> {
+        self.client
+            .notify_outputs_written(proto::NotifyOutputsWrittenRequest {
+                hash,
+                output_globs,
+                output_exclusion_globs,
+            })
+            .await?;
+
+        Ok(())
+    }
+}
+
+impl DaemonClient<()> {
+    /// Augment the client with the connect settings, allowing it to be
+    /// restarted.
+    pub fn with_connect_settings(
+        self,
+        connect_settings: DaemonConnector,
+    ) -> DaemonClient<DaemonConnector> {
+        DaemonClient {
+            client: self.client,
+            connect_settings,
+        }
+    }
+}
+
+impl DaemonClient<DaemonConnector> {
+    /// Stops the daemon, closes the connection, and opens a new connection.
+    pub async fn restart(self) -> Result<DaemonClient<DaemonConnector>, DaemonError> {
+        self.stop().await?.connect().await
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum DaemonError {
+    #[error("Failed to connect to daemon")]
+    Connection,
+    #[error("Daemon version mismatch")]
+    VersionMismatch,
+    #[error("could not connect: {0}")]
+    GrpcTransport(#[from] tonic::transport::Error),
+    #[error("could not fork")]
+    Fork,
+    #[error("could not connect: {0}")]
+    GrpcFailure(tonic::Code),
+    #[error("missing response")]
+    MissingResponse,
+    #[error("could not read pid file")]
+    PidFile,
+    #[error("could not connect: {0}")]
+    Timeout(#[from] tokio::time::error::Elapsed),
+    #[error("daemon is not running and will not be started")]
+    NotRunning,
+}
```

Comment:
Do these show up in our UI? If so I'd love to make them a bit more accessible.

Reply:
some of them will bubble up. I am going to clean up the error paths and expose the underlying ones a little better